include/alpaka/block/shared/st/detail/BlockSharedMemStMemberImpl.hpp

```diff
@@ -0,0 +1,168 @@
+                uint32_t off = 0;
+
+                // Iterate over allocated data only
+                while(off < m_allocdBytes)
+                {
+                    // Adjust offset to be aligned
+                    uint32_t const alignedMetaDataOffset
+                        = varChunkEnd<MetaData>(off) - static_cast<uint32_t>(sizeof(MetaData));
+                    ALPAKA_ASSERT_OFFLOAD(
+                        (alignedMetaDataOffset + static_cast<uint32_t>(sizeof(MetaData))) <= m_allocdBytes);
+                    MetaData* metaDataPtr = reinterpret_cast<MetaData*>(m_mem + alignedMetaDataOffset);
+                    off = metaDataPtr->offset;
+
+                    if(metaDataPtr->id == id)
+                        break;
+                }
+
+                // Variable not found.
+                if(off >= m_allocdBytes)
+                    return nullptr;
+
+                return reinterpret_cast<T*>(&m_mem[off - sizeof(T)]);
+            }
+
+            //! Get last allocated variable.
+            template<typename T>
+            T* getLatestVarPtr() const
+            {
+                return reinterpret_cast<T*>(&m_mem[m_allocdBytes - sizeof(T)]);
+            }
+
+        private:
+#if BOOST_COMP_GNUC
+#    pragma GCC diagnostic pop
+#endif
+
+            //! Byte offset to the end of the memory chunk
+            //!
+            //! Calculate bytes required to store a type with a aligned starting offset.
+            //! Start offset to the origin of the user data chunk can be calculated with `result - sizeof(T)`.
+            //! The padding is always before the origin of the user data chunk and can be zero byte.
+            //!
+            //! \tparam T type should fit into the chunk
+            //! \param byteOffset Current byte offset.
+            //! \result Byte offset to the end of the data chunk, relative to m_mem..
+            template<typename T>
+            std::uint32_t varChunkEnd(uint32_t byteOffset) const
+            {
+                static_assert(
+                    core::vectorization::defaultAlignment >= alignof(T),
+                    "Unable to get block shared static memory for types with alignment higher than defaultAlignment!");
+                constexpr std::uint32_t align = static_cast<std::uint32_t>(std::max(TMinDataAlignBytes, alignof(T)));
+                return (byteOffset / align + (byteOffset % align > 0u)) * align + static_cast<uint32_t>(sizeof(T));
+            }
+
+            //! Offset in bytes relative to m_mem to next free data area.
+            //! The last aligned before the free area is always a meta data header.
+            mutable std::uint32_t m_allocdBytes = 0u;
+
+            //! Memory layout
+            //! |Header|Padding|Variable|Padding|Header|....uninitialized Data ....
+            //! Size of padding can be zero if data after padding is already aligned.
+            mutable uint8_t* m_mem;
```

Comment:
I think `m_mem` does not need to be `mutable`.

Reply:
It should be mutable because the accelerator is inheriting from this class and is a constant object in all kernels.

Reply:
`m_mem` is never modified. It does not need to be mutable to point to mutable memory. The actual underlying memory must be mutable, but this is [here](https://github.com/alpaka-group/alpaka/blob/develop/include/alpaka/block/shared/dyn/BlockSharedMemDynMember.hpp#L104).

Reply:
@psychocoderHPC you could straight-ip put it down as:
```c++
uint8_t* const m_mem;
```

Reply:
I considered adding `const` but I do not know if inheriting classes need to be movable, which no longer works with const data members.

Reply:
E.g.:
https://github.com/alpaka-group/alpaka/blob/ed3c123b22684b9a342506d4d46a03a8ac6bd9bf/include/alpaka/acc/AccCpuSerial.hpp#L112
... no.

Reply:
Fine! So `uint8_t* const m_mem;` is probably a good choice then.

Reply:
right I mixed the pointer with the value to store how many bytes we already allocated.
I will change it.